src/dht_proxy_client.cpp

```diff
@@ -0,0 +1,587 @@
+            ipv6Address = ips.substr(0, ipv4And6Separator - 1);
+        }
+    } else {
+        auto endIp = public_ip.find_last_of(':');
+        port = public_ip.substr(endIp + 1);
+        ipv4Address = public_ip.substr(0, endIp - 1);
+    }
+    switch (family)
+    {
+    case AF_INET:
+        return DhtRunner::getAddrInfo(ipv4Address, port);
+    case AF_INET6:
+        return DhtRunner::getAddrInfo(ipv6Address, port);
+    default:
+        return {};
+    }
+}
+
+size_t
+DhtProxyClient::listen(const InfoHash& key, GetCallback cb, Value::Filter&& filter, Where&& where)
+{
+    restbed::Uri uri(HTTP_PROTO + serverHost_ + "/" + key.toString());
+    auto req = std::make_shared<restbed::Request>(uri);
+    req->set_method("LISTEN");
+
+    Query query {{}, where};
+    auto filterChain = filter.chain(query.where.getFilter());
+
+    Listener l;
+    ++listener_token_;
+    l.key = key.toString();
+    l.token = listener_token_;
+    l.req = req;
+    l.cb = cb;
+    l.filterChain = std::move(filterChain);
+    l.thread = std::move(std::thread([=]()
+        {
+            auto settings = std::make_shared<restbed::Settings>();
+            std::chrono::milliseconds timeout(std::numeric_limits<int>::max());
+            settings->set_connection_timeout(timeout); // Avoid the client to close the socket after 5 seconds.
+
+            restbed::Http::async(req,
+                [this, filterChain, cb](const std::shared_ptr<restbed::Request>& req,
+                                        const std::shared_ptr<restbed::Response>& reply) {
+                auto code = reply->get_status_code();
+
+                if (code == 200) {
+                    try {
+                        while (restbed::Http::is_open(req)) {
+                            restbed::Http::fetch("\n", reply);
+                            std::string body;
+                            reply->get_body(body);
+                            reply->set_body(""); // Reset the body for the next fetch
+
+                            Json::Value json;
+                            Json::Reader reader;
+                            if (reader.parse(body, json)) {
+                                auto value = std::make_shared<Value>(json);
+                                if ((not filterChain or filterChain(*value)) && cb)  {
+                                    lockCallbacks.lock();
+                                    callbacks_.emplace_back([=](){
+                                        cb({value});
+                                    });
+                                    lockCallbacks.unlock();
+                                }
+                            }
+                        }
+                    } catch (std::runtime_error&) {
+                        // NOTE: Http::close() can occurs here. Ignore this.
+                    }
+
+                } else {
+                    this->statusIpv4_ = NodeStatus::Disconnected;
```

Comment:
why a different behavior from get/put ?

Reply:
Old code not replaced in one case. Done.